### PR TITLE
Add size threshold for few oneDNN operators

### PR DESCRIPTION
**Pull request by bgawrych**
## Description ##
Some oneDNN operators are not optimal choice when working on dynamic shaped input like in GPT-2. This change add size threshold to determine whether to generate JIT kernel or run with native one

Benchmarks on model from gluon-nlp (AWS C6i.x12large) with following OMP settings (OMP_PROC_BIND=true OMP_PLACES=sockets OMP_NUM_THREADS=24)

image

### CODE REVIEW

**Comment by bgawrych** [comment]
@mxnet-bot run ci [all]

File: src/operator/nn/dnnl/dnnl_binary.cc [hunk 1]
```diff
@@ -65,7 +65,11 @@ void DNNLBinaryOpFwd::Execute(const std::vector<NDArray>& inputs,
 
 // Support for https://oneapi-src.github.io/oneDNN/v2.6/dev_guide_binary.html
 bool SupportDNNLBinary(const std::vector<NDArray>& inputs) {
-  return SupportDNNL<DNNLTypeMode::FloatTypes>(inputs[1]) &&
+  // threshold value selected experimentally basing on performance results - PR-21106
+  const size_t optimal_size_threshold = 2 << 13;
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_binary.cc:69]:
```suggestion
  constexpr size_t optimal_size_threshold = 2 << 13;
```

File: src/operator/nn/dnnl/dnnl_masked_softmax.cc [hunk 2]
```diff
@@ -31,12 +31,16 @@ namespace op {
...
+  const auto data = inputs[0];
   const auto mask = inputs[1];
   SoftmaxParam softmax_param;
   softmax_param.axis        = param.axis;
-  softmax_param.dtype       = inputs[0].dtype();
+  softmax_param.dtype       = data.dtype();
   softmax_param.temperature = param.temperature;
-  return mask.dtype() == mshadow::kBool && SupportDNNLSoftmax(softmax_param, inputs[0]);
+  // threshold value selected experimentally basing on performance results - PR-21106
+  const size_t optimal_size_threshold = 2 << 13;
```
Comment (bartekkuncer) [inline src/operator/nn/dnnl/dnnl_masked_softmax.cc:41]:
```suggestion
  constexpr size_t optimal_size_threshold = 2 << 13;
```

**Comment by bgawrych** [comment]
@mxnet-bot run ci [edge, unix-cpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-cpu, windows-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu, centos-gpu]

**Comment by bgawrych** [comment]
@mxnet-bot run ci [unix-gpu, centos-gpu]

**Review by bartekkuncer — APPROVED** [review]
LGTM

**Pull request merged**